java-frontend/src/main/java/org/sonar/java/se/checks/StreamPipeline.java

```diff
@@ -0,0 +1,182 @@
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.ReturnStatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+import javax.annotation.Nullable;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+
+public class StreamPipeline extends SECheck {
+
+  private StreamConsumedCheck streamConsumedCheck;
+
+  public StreamPipeline(@Nullable StreamConsumedCheck streamConsumedCheck) {
+    this.streamConsumedCheck = streamConsumedCheck;
+  }
+
+  public enum StreamPipelineConstraint implements Constraint {
+    CONSUMED, NOT_CONSUMED
+  }
+
+  private static final Set<String> STREAM_TYPES = ImmutableSet.of("java.util.stream.Stream", "java.util.stream.IntStream", "java.util.stream.LongStream",
+    "java.util.stream.DoubleStream");
+  private static final MethodMatcherCollection TERMINAL_OPERATIONS = MethodMatcherCollection.create();
+  static {
+    List<String> terminalMethods = ImmutableList.of("forEach", "forEachOrdered", "toArray", "collect", "reduce", "findAny", "findFirst", "count", "min", "max", "anyMatch",
+      "allMatch", "noneMatch", "average", "summaryStatistics");
+
+    STREAM_TYPES.forEach(streamType -> terminalMethods.forEach(method ->
+      TERMINAL_OPERATIONS.add(MethodMatcher.create().typeDefinition(streamType).name(method).withAnyParameters())));
+  }
+  private static final MethodMatcherCollection INTERMEDIATE_OPERATIONS = MethodMatcherCollection.create();
+  static {
+    List<String> intermediateMethods = ImmutableList.of("filter", "map", "mapToInt", "mapToLong", "mapToDouble", "flatMap", "flatMapToInt", "flatMapToLong", "flatMapToDouble",
+      "flatMapToObj", "distinct", "sorted", "peek", "limit", "mapToObj", "skip");
+    STREAM_TYPES.forEach(streamType -> intermediateMethods.forEach(method ->
+      INTERMEDIATE_OPERATIONS.add(MethodMatcher.create().typeDefinition(streamType).name(method).withAnyParameters())));
+  }
+
+  @Override
+  public ProgramState checkPreStatement(CheckerContext context, Tree syntaxNode) {
+    if (syntaxNode.is(Tree.Kind.METHOD_INVOCATION)) {
+      return handleMethodInvocation(context, syntaxNode);
+    }
+    return context.getState();
+  }
+
+  private ProgramState handleMethodInvocation(CheckerContext context, Tree syntaxNode) {
+    MethodInvocationTree mit = (MethodInvocationTree) syntaxNode;
+    ProgramState programState = context.getState();
+    programState = removeConstraintOnArgs(programState, mit);
+    SymbolicValue invocationTarget = invocationTarget(programState, mit);
+    if ((isIntermediateOperation(mit) || isTerminalOperation(mit))
+        && isPipelineConsumed(programState, invocationTarget)
+        && streamConsumedCheck != null) {
+      streamConsumedCheck.reportIssue(syntaxNode, "Refactor this code so that this consumed stream pipeline is not reused.", flow(invocationTarget, context.getNode()));
+      return null;
```

Comment:
because this stop the symbolic execution engine and thus contribute to the exploration of the EG this should always be executed and reported only if the check is not null. Otherwise we might have issue appearing if we deactivate stream rule and disappearing when activating it.